utilities/transactions/lock/range/range_tree/range_tree_lock_manager.cc

```diff
@@ -0,0 +1,572 @@
+
+  if (mutex_factory)
+    use_factory = mutex_factory;
+  else
+    use_factory.reset(new TransactionDBMutexFactoryImpl());
+
+  return new RangeTreeLockManager(use_factory);
+}
+
+static const char SUFFIX_INFIMUM = 0x0;
+static const char SUFFIX_SUPREMUM = 0x1;
+
+void serialize_endpoint(const Endpoint& endp, std::string* buf) {
+  buf->push_back(endp.inf_suffix ? SUFFIX_SUPREMUM : SUFFIX_INFIMUM);
+  buf->append(endp.slice.data(), endp.slice.size());
+}
+
+void deserialize_endpoint(const DBT* dbt, EndpointWithString* endp) {
+  assert(dbt->size >= 1);
+  const char* dbt_data = (const char*)dbt->data;
+  char suffix = dbt_data[0];
+  assert(suffix == SUFFIX_INFIMUM || suffix == SUFFIX_SUPREMUM);
+  endp->inf_suffix = (suffix == SUFFIX_SUPREMUM);
+  endp->slice.assign(dbt_data + 1, dbt->size - 1);
+}
+
+// Get a range lock on [start_key; end_key] range
+Status RangeTreeLockManager::TryLock(PessimisticTransaction* txn,
+                                     uint32_t column_family_id,
+                                     const Endpoint& start_endp,
+                                     const Endpoint& end_endp, Env*,
+                                     bool exclusive) {
+  toku::lock_request request;
+  request.create(mutex_factory_);
+  DBT start_key_dbt, end_key_dbt;
+
+  TEST_SYNC_POINT("RangeTreeLockManager::TryRangeLock:enter");
+  std::string start_key;
+  std::string end_key;
+  serialize_endpoint(start_endp, &start_key);
+  serialize_endpoint(end_endp, &end_key);
+
+  toku_fill_dbt(&start_key_dbt, start_key.data(), start_key.size());
+  toku_fill_dbt(&end_key_dbt, end_key.data(), end_key.size());
+
+  // Use the txnid as "extra" in the lock_request. Then, KillLockWait()
+  // will be able to use kill_waiter(txn_id) to kill the wait if needed
+  // TODO: KillLockWait is gone and we are no longer using
+  // locktree::kill_waiter call. Do we need this anymore?
+  TransactionID wait_txn_id = txn->GetID();
+
+  auto lt = get_locktree_by_cfid(column_family_id);
+
+  request.set(lt, (TXNID)txn, &start_key_dbt, &end_key_dbt,
+              exclusive ? toku::lock_request::WRITE : toku::lock_request::READ,
+              false /* not a big txn */, (void*)wait_txn_id);
+
+  // This is for "periodically wake up and check if the wait is killed" feature
+  // which we are not using.
+  uint64_t killed_time_msec = 0;
+  uint64_t wait_time_msec = txn->GetLockTimeout();
+
+  // convert microseconds to milliseconds
+  if (wait_time_msec != (uint64_t)-1)
+    wait_time_msec = (wait_time_msec + 500) / 1000;
+
+  std::vector<RangeDeadlockInfo> di_path;
+  request.m_deadlock_cb = [&](TXNID txnid, bool is_exclusive,
+                              const DBT* start_dbt, const DBT* end_dbt) {
+    EndpointWithString start;
+    EndpointWithString end;
+    deserialize_endpoint(start_dbt, &start);
+    deserialize_endpoint(end_dbt, &end);
+
+    di_path.push_back({((PessimisticTransaction*)txnid)->GetID(),
```

Comment:
can we use `emplace_back` and get rid of the `{}` construction inside?

Reply:
This doesn't seem to work, as `RangeDeadlockInfo` doesn't have a constructor. 

It's the same with point-based `DeadlockInfo`.  It is also filled with `push_back`, not `emplace_back`:
https://github.com/facebook/rocksdb/blob/master/utilities/transactions/lock/point/point_lock_manager.cc#L430